### ISSUE TITLE
setting uri using service external IP

### ISSUE THREAD

**Issue by oxeye-schmil** [aws-controllers-k8s/community#1005]
Hi

According to https://aws.amazon.com/blogs/containers/integrate-amazon-api-gateway-with-amazon-eks/
we should aws cli to set the apigateway uri
In my helm chart I define a service that uses aws lb
I also define an apigateway and would like to use the service fqdn (erternal-ip) to set the uri
Any idea how to do it in the same chart?

### service definition

`kind: Service
metadata:
  name: {{.Release.Name}}-{{.Chart.Name}}-collector
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-type: external
    service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: ip
spec:
  ports:
    - port: {{ .Values.collector.config.ports.http }}
      targetPort: {{ .Values.collector.config.ports.http }}
      protocol: TCP
  type: LoadBalancer
  selector:
    app: {{.Release.Name}}-{{.Chart.Name}}-collector
`

### api gateway v2 definition

apiVersion: apigatewayv2.services.k8s.aws/v1alpha1
kind: API
metadata:
  name: saas-apigwv2-nlb
spec:
  body: '{
            "openapi": "3.0.1",
              "info": {
                "title": "saas-apigwv2-nlb",
                "version": "v1"
              },
              "paths": {
              "/\$default": {
                "x-amazon-apigateway-any-method" : {
                "isDefaultRoute" : true,
                "x-amazon-apigateway-integration" : {
                "payloadFormatVersion" : "1.0",
                "connectionId" : " {{ required "vpc ID required" .Values.apiGatewayV2.vpcID }} ",
                "type" : "http_proxy",
                "httpMethod" : "POST",
                "uri" : "$(aws elbv2 describe-listeners --load-balancer-arn $(aws elbv2 describe-load-balancers --region $AGW_AWS_REGION --query "LoadBalancers[?contains(DNSName, '$(kubectl get service authorservice -o jsonpath="{.status.loadBalancer.ingress[].hostname}")')].LoadBalancerArn" --output text) --region $AGW_AWS_REGION --query "Listeners[0].ListenerArn" --output text)",
                "connectionType" : "VPC_LINK"
                  }
                }
              },
              "components": {}
          }'

**Comment by RedbackThomson** [comment]
Hey @oxeye-schmil,
This seems like an interesting proposition, and I can totally see value in a dynamic interaction between the two controllers. From what I understand, you would essentially like to inject the URI created by the ELB controller into the body of the APIGW `API` object within the same chart?

ACK controllers don't currently support something like that. I imagine this would be supported in the form of `ConfigMap` values, or something of that sort. For the moment, we expect that each spec field is statically defined, however. One suggestion I might make is that you install the ALB controller and your service annotations in one chart, and the ACK APIGWv2 controller and `API` in another chart. You can then set the `body` field of your `API` object from a Helm chart value, which can be modified at apply-time (eg. `helm install --set body.uri=$(...) ...`).

Hope that answers your question.

**Comment by oxeye-schmil** [comment]
I did separate the charts for the moment
but it causes dependency between charts that causes devops issues
Can we tag it as a feature request?

**Comment by vijtrip2** [comment]
@oxeye-schmil , Please correct me if my understanding below is incorrect.

You want to create a helm chart which has a k8s service(LoadBalancerType) and this k8s service's external IP will be used as uri in API resource.
The issue that you may be seeing are that API Gateway resource is not waiting for k8s service to be created and you'd like to have a dependency relation between them.

If above is correct, then we have a similar [issue#545](https://github.com/aws-controllers-k8s/community/issues/545) that we will work on in the future to handle resource relationship.
But Note that resource relationship being handled there will be between different resources managed by same ACK controller. Ex: VpcId and API resource.

So I am not sure if we can take this up as a feature request to support relationship between resources not managed by same controller. And this is completely from a technical viewpoint because to establish relationship you need to add informers & watchers in controller.

**Comment by RedbackThomson** [comment]
I will also point you to the following issues:
https://github.com/aws-controllers-k8s/community/issues/20
https://github.com/aws-controllers-k8s/community/issues/740

**Comment by vijtrip2** [comment]
> By the way is possible to have the API resource run in "default" stage wihtout creating a stage resource?

No, you'll need to create the stage resource for invocation. It is possible to create the stage with name "$default", which will allow you to skip stage-name in invocation url.
